Mathlib/SetTheory/Ordinal/Topology.lean

```diff
@@ -234,4 +234,126 @@ theorem enumOrd_isNormal_iff_isClosed (hs : s.Unbounded (· < ·)) :
     exact le_bsup.{u, u} _ _ (ha.2 _ hba)
 #align ordinal.enum_ord_is_normal_iff_is_closed Ordinal.enumOrd_isNormal_iff_isClosed
 
+open Set Filter
+
+/-- A positive ordinal is an accumulation point of a set of ordinals if there
+are elements in the set arbitrarily close to the ordinal from below. -/
+def IsAcc (o : Ordinal) (S : Set Ordinal) : Prop :=
+  o ≠ 0 ∧ ∀ p < o, ∃ s ∈ S, p < s ∧ s < o
+
+/-- A set of ordinals is closed in an ordinal if it contains all of
+its accumulation points below the ordinal. -/
+def IsClosed (S : Set Ordinal) (o : Ordinal) : Prop :=
+  ∀ p < o, IsAcc p S → p ∈ S
+
+theorem IsAcc.subset {o : Ordinal} {S T : Set Ordinal} (h : S ⊆ T) (ho : o.IsAcc S) :
+    o.IsAcc T := ⟨ho.1, fun p plto ↦ (ho.2 p plto).casesOn fun s hs ↦ ⟨s, h hs.1, hs.2⟩⟩
+
+theorem IsAcc.isLimit {o : Ordinal} {U : Set Ordinal} (h : o.IsAcc U) : IsLimit o := by
+  refine' isLimit_of_not_succ_of_ne_zero (fun ⟨x, hx⟩ ↦ _) h.1
+  rcases h.2 x (lt_of_lt_of_le (lt_succ x) hx.symm.le) with ⟨p, hp⟩
+  exact (hx.symm ▸ (succ_le_iff.mpr hp.2.1)).not_lt hp.2.2
+
+theorem IsAcc.inter_Ioo_nonempty {o : Ordinal} {U : Set Ordinal} (hU : o.IsAcc U)
+    {p : Ordinal} (hp : p < o) : (U ∩ Ioo p o).Nonempty := by
+  rcases hU.2 p hp with ⟨q, hq⟩
+  exact ⟨q, hq.1, hq.2.1, hq.2.2⟩
+
+theorem isClosed_zero (S : Set Ordinal) : IsClosed S 0 := fun _ h ↦
+  False.elim <| (Ordinal.zero_le _).not_lt h
+
+theorem isAcc_of_accPt {o p : Ordinal} (A : Set Ordinal) (hpos : p ≠ 0) (hp : p < o)
+    (h : AccPt ⟨p, hp⟩ (𝓟 {x : Iio o | x.1 ∈ A})) : IsAcc p A := by
+  use hpos
+  rw [accPt_iff_nhds] at h
+  intro q qltp
+  by_cases h' : Order.succ p = o
+  · let I : Set (Iio o) := Ioi ⟨q, qltp.trans hp⟩
+    have Inh : I ∈ nhds ⟨p, hp⟩ := IsOpen.mem_nhds isOpen_Ioi qltp
+    rcases h I Inh with ⟨r, ⟨rmemI, rmemA⟩, rnep⟩
+    use r, rmemA, rmemI
+    have : r ≤ p := Order.lt_succ_iff.mp (h' ▸ r.2)
+    exact this.lt_or_eq.resolve_right <| Subtype.coe_ne_coe.mpr rnep
```

Comment:
If you move the `have` block one line earlier, you can combine `use` and `exact` into a single `exact`.